sympy/solvers/solvers.py

```diff
@@ -863,64 +857,141 @@ def _sympified_list(w):
 
+    # alternate exits
 
-    # preprocess equation(s)
-    ###########################################################################
-    for i, fi in enumerate(f):
-        if isinstance(fi, (Equality, Unequality)):
-            if 'ImmutableDenseMatrix' in [type(a).__name__ for a in fi.args]:
-                fi = fi.lhs - fi.rhs
-            else:
-                L, R = fi.args
-                if isinstance(R, BooleanAtom):
-                    L, R = R, L
-                if isinstance(L, BooleanAtom):
-                    if isinstance(fi, Unequality):
-                        L = ~L
-                    if R.is_Relational:
-                        fi = ~R if L is S.false else R
-                    elif R.is_Symbol:
-                        return L
-                    elif R.is_Boolean and (~R).is_Symbol:
-                        return ~L
-                    else:
-                        raise NotImplementedError(filldedent('''
-                            Unanticipated argument of Eq when other arg
-                            is True or False.
-                        '''))
+    def is_eq_bool(eq):
+        if isinstance(eq, (Equality, Unequality)):
+            return any(side in (S.false, S.true) for side in eq.args)
+
+    if any(is_eq_bool(fi) for fi in f):
+        for i, fi in enumerate(f):
+            if not isinstance(fi, (Equality, Unequality)):
+                continue
+            args = fi.args
+            if args[1] in (S.true, S.false):
+                args = args[1], args[0]
+            L, R = args
+            if L in (S.false, S.true):
+                if isinstance(fi, Unequality):
+                    L = ~L
+                if R.is_Relational:
+                    pass
+                elif R.is_Symbol:
+                    return L
+                elif R.is_Boolean and (~R).is_Symbol:
+                    return ~L
                 else:
-                    fi = fi.rewrite(Add, evaluate=False)
-            f[i] = fi
+                    raise NotImplementedError(filldedent('''
+                        Unanticipated argument of Eq when other arg
+                        is True or False.
+                    '''))
 
-        if fi.is_Relational:
-            return reduce_inequalities(f, symbols=symbols)
+    def invert_eq_bool(fi):
+        if isinstance(fi, (Equality, Unequality)):
+            lhs, rhs = fi.args
+            unequal = S.true if isinstance(fi, Unequality) else S.false
+            if lhs in (S.false, S.true) and rhs.is_Relational:
+                return ~rhs if lhs is unequal else rhs
+            if rhs in (S.false, S.true) and lhs.is_Relational:
+                return ~lhs if rhs is unequal else lhs
+        return fi
+
+    f_relational = [invert_eq_bool(fi) for fi in f]
+
+    def is_relational(fi):
+        return fi.is_Relational and not isinstance(fi, Equality)
+
+    if any(is_relational(fi) for fi in f_relational):
+        f = [invert_eq_bool(fi) for fi in f_relational]
```

Comment:
Can we move everything in this `if` block to a separate function?

Since these things have separate logic from the rest of solve we should try to separate them in the code to make that clear.

Reply:
Moving it to a new function is breaking things

Reply:
Turns out, I was wrong. These can be moved to a new function without breaking anything.